conceptnet5/vectors/query.py

```diff
@@ -129,47 +131,83 @@ def expand_terms(self, terms, limit_per_term=10, include_neighbors=True):
         """
         self.load()
         expanded = terms[:]
+        prefix_weight = 0.01
         for term, weight in terms:
-            if include_neighbors and term not in self.frame.index and self.finder is not None:
-                for edge in self.finder.lookup(term, limit=limit_per_term):
-                    if field_match(edge['start']['term'], term) and not field_match(
-                            edge['end']['term'], term):
-                        neighbor = edge['end']['term']
-                    elif field_match(edge['end']['term'], term) and not field_match(
-                            edge['start']['term'], term):
-                        neighbor = edge['start']['term']
-                    else:
-                        continue
-                    # TODO: explain this formula
-                    neighbor_weight = weight * min(10, edge['weight']) * 0.01
-                    expanded.append((neighbor, neighbor_weight))
-
-                prefix_weight = 0.01
+            if handle_oov and term not in self.frame.index and self.finder is not None:
+
+                # Get neighbors
+                neighbors = self.get_neighbors(term, limit_per_term, weight)
+                expanded.extend(neighbors)
+
+                # Get englishified version of the term
                 if get_language(term) != 'en':
                     englishified = '/c/en/' + split_uri(term)[2]
                     expanded.append((englishified, prefix_weight))
 
-                while term:
-                    # Skip excessively general lookups, for either an entire
-                    # language, or all terms starting with a single
-                    # non-ideographic letter
-                    if term.endswith('/') or (term[-2] == '/' and term[-1] < chr(0x3000)):
-                        break
-                    prefixed = self.terms_with_prefix(term)
-                    if prefixed:
-                        n_prefixed = len(prefixed)
-                        for prefixed_term in prefixed:
-                            expanded.append((prefixed_term, prefix_weight / n_prefixed))
-                        break
-                    term = term[:-1]
+                # Get matching prefixes
+                prefixes = self.prefix_matching(term, prefix_weight)
+                expanded.extend(prefixes)
+
+                # Get subwords
+                # subwords = self.get_subwords(term)
+                # expanded.extend([(subword, prefix_weight) for subword in subwords])
```

Comment:
Is this meant to be commented out? Is it not ready yet?

One thing I'm worried about: loading a Morfessor model from its text format has unreasonably high latency for use cases such as the ConceptNet API. But the only binary format it has is the pickle format, which bit rots.

The existence of Jieba shows that a subword model _can_ be loaded from a binary format in about half a second; I think we're going to need something like that.

Reply:
This was commented out because on average our results were hurt by adding the subwords. I think the solution here may be to add `use_subwords` as a param to `expand_terms`.

Good point about Jieba, will take a look.